### PR TITLE
[TACHYON-993] Implement a better exists method

### PR DESCRIPTION
**Pull request by manugoyal**
https://tachyon.atlassian.net/browse/TACHYON-993

Currently, to check if a file exists, users have to call getInfo on the
file, and catch if it throws an exception. Instead, we changed
TachyonFileSystem.open, so that when a user opens a file, we return null
if the file was not found. Thus users can check if a file exists by
trying to open it.

### CODE REVIEW

**Comment by manugoyal** [comment]
Jenkins, test this please

File: clients/unshaded/src/main/java/tachyon/client/TachyonFS.java [hunk 1]
```diff
@@ -400,10 +401,9 @@ public synchronized boolean delete(long fileId, TachyonURI path, boolean recursi
       throws IOException {
     validateUri(path);
     if (fileId == -1) {
-      try {
-        fileId = mFSMasterClient.getFileId(path.getPath());
-      } catch (InvalidPathException e) {
-        throw new IOException(e);
+      fileId = mFSMasterClient.getFileId(path.getPath());
+      if (fileId == -1) {
+        throw new IOException(ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(path.getPath()));
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFS.java]: Maybe its cleaner to move this logic to a utils function like `validateFileId`?

**Comment by calvinjia** [comment]
@manugoyal Thanks for the patch, could you also add the change in `TachyonFileSystem` to implement `exists(): boolean` using `open`?

File: clients/unshaded/src/main/java/tachyon/client/TachyonFS.java (hunk 1 above)
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFS.java]: +1 or maybe called `checkFileId`

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 2]
```diff
@@ -251,7 +251,7 @@ private void copyToLocalWithBytes(int bytes) throws IOException {
   @Test
   public void countNotExistTest() throws IOException {
     int ret = mFsShell.run(new String[] {"count", "/NotExistFile"});
-    Assert.assertEquals("Could not find path: /NotExistFile\n", mOutput.toString());
+    Assert.assertEquals("Path /NotExistFile does not exist\n", mOutput.toString());
```
Comment (apc999) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: `ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage("/NotExistFile")`?

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 3]
```diff
@@ -664,7 +664,7 @@ public void duTest() throws IOException {
     String expected = "";
     // du a non-existing file
     mFsShell.run(new String[] {"du", "/testRoot/noneExisting"});
-    expected += "Could not find path: /testRoot/noneExisting\n";
+    expected += "Path /testRoot/noneExisting does not exist\n";
```
Comment (apc999) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: ditto

File: servers/src/test/java/tachyon/master/file/FileSystemMasterTest.java [hunk 4]
```diff
@@ -165,7 +159,7 @@ public void renameTest() throws Exception {
   @Test
   public void renameUnderNonexistingDir() throws Exception {
     mThrown.expect(InvalidPathException.class);
-    mThrown.expectMessage("Could not find path: /nested/test");
+    mThrown.expectMessage("Path /nested/test does not exist");
```
Comment (apc999) [inline servers/src/test/java/tachyon/master/file/FileSystemMasterTest.java]: using `PATH_DOES_NOT_EXIST` to format the expected Message?

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java [hunk 5]
```diff
@@ -202,7 +204,7 @@ public void setPin(TachyonFile file, boolean pinned) throws FileDoesNotExistExce
 
   @Override
   public boolean rename(TachyonFile src, TachyonURI dst) throws FileDoesNotExistException,
-      IOException {
+          IOException, FileAlreadyExistException, InvalidPathException {
```
Comment (apc999) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: indent seems off for line 207

File: common/src/main/java/tachyon/client/FileSystemMasterClient.java [hunk 6]
```diff
@@ -354,11 +352,13 @@ public synchronized boolean deleteFile(long fileId, boolean recursive) throws IO
...
    * @return whether operation succeeded or not
-   * @throws FileDoesNotExistException if the file does not exist
+   * @throws FileDoesNotExistException if the source file does not exist
+   * @throws FileAlreadyExistException if there is already a file at the destination path
+   * @throws InvalidPathException if the destination path is invalid
    * @throws IOException if an I/O error occurs
    */
   public synchronized boolean renameFile(long fileId, String dstPath) throws IOException,
-      FileDoesNotExistException {
+          FileDoesNotExistException, FileAlreadyExistException, InvalidPathException {
```
Comment (apc999) [inline common/src/main/java/tachyon/client/FileSystemMasterClient.java]: indent seems off here too.

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 7]
```diff
@@ -313,12 +311,16 @@ public boolean isDirectory(long id) {
...
    */
-  public long getFileId(TachyonURI path) throws InvalidPathException {
+  public long getFileId(TachyonURI path) {
     synchronized (mInodeTree) {
-      Inode inode = mInodeTree.getInodeByPath(path);
+      Inode inode;
+      try {
+        inode = mInodeTree.getInodeByPath(path);
+      } catch (InvalidPathException e) {
+        return -1;
```
Comment (apc999) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: Right now in theory, -1 is still a valid fileId that a block master may assign. Though I think the risk of running into this issue is low, I think it is necessary to have a TODO here to make sure client and master agrees on some impossible file ids.

**Comment by apc999** [comment]
Thanks @manugoyal , I left some comments.

**Comment by manugoyal** [comment]
Jenkins, test this please

File: servers/src/main/java/tachyon/master/file/FileSystemMaster.java [hunk 8]
```diff
@@ -362,12 +361,17 @@ public boolean isDirectory(long id) {
    * Returns the file id for a given path. Called via RPC, as well as internal masters.
    *
    * @param path the path to get the file id for
-   * @return the file id for a given path
-   * @throws InvalidPathException
+   * @return the file id for a given path, or -1 if there is no file at that path
    */
-  public long getFileId(TachyonURI path) throws InvalidPathException {
+  public long getFileId(TachyonURI path) {
+    // TODO(manugoyal) make sure client and master agree that -1 is an invalid file id
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/FileSystemMaster.java]: To enforce this, you can define a constant in IdUtils, and let both client and master reference it.

**Comment by yupeng9** [comment]
Thanks for the improvement. I left one comment

**Comment by manugoyal** [comment]
Note that we don't actually do anything other than log a warning right now if you're creating a file id of -1, but it's now made explicit that -1 is invalid.

**Comment by yupeng9** [comment]
LGTM. Can you resolve the conflicts?

File: clients/unshaded/src/main/java/tachyon/client/TachyonFS.java [hunk 9]
```diff
@@ -538,6 +533,23 @@ public synchronized TachyonFile getFile(TachyonURI path, boolean useCachedMetada
...
 
   /**
+   * If the given file id is -1, tries to fetch the file id for the given path
+   * @param fileId
+   * @param path the path to search for if fileId is -1
+   * @return the original fileId if it wasn't -1, or a new fileId corresponding to the given path
+   * @throws IOException if the given fileId is -1 and no fileId was found for the given path
+   */
+  private long getValidFileId(long fileId, String path) throws IOException {
+    if (fileId == -1) {
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/TachyonFS.java]: Shall we also use the constant defined in IdUtils?

**Comment by yupeng9** [comment]
LGTM. Thanks for the improvement!

**Pull request merged**